pkg/model/awsmodel/oidc_provider.go

```diff
@@ -35,43 +43,163 @@ type OIDCProviderBuilder struct {
+	ClaimsSupported       []string `json:"claims_supported"`
+}
+
 var _ fi.ModelBuilder = &OIDCProviderBuilder{}
 
 const (
 	defaultAudience = "amazonaws.com"
 )
 
 func (b *OIDCProviderBuilder) Build(c *fi.ModelBuilderContext) error {
-	var thumbprints []fi.Resource
-	var issuerURL string
 
-	if featureflag.PublicJWKS.Enabled() {
-		serviceAccountIssuer := iam.ServiceAccountIssuer(b.ClusterName(), &b.Cluster.Spec)
-		issuerURL = serviceAccountIssuer
+	if !featureflag.PublicJWKS.Enabled() {
+		return nil
+	}
 
-		caTaskObject, found := c.Tasks["Keypair/ca"]
-		if !found {
-			return fmt.Errorf("keypair/ca task not found")
-		}
+	serviceAccountIssuer, err := iam.ServiceAccountIssuer(&b.Cluster.Spec)
+	if err != nil {
+		return err
+	}
 
-		caTask := caTaskObject.(*fitasks.Keypair)
-		fingerprint := caTask.CertificateSHA1Fingerprint()
+	signingKeyTaskObject, found := c.Tasks["Keypair/master"]
+	if !found {
+		return fmt.Errorf("keypair/master task not found")
+	}
 
-		thumbprints = []fi.Resource{fingerprint}
+	fingerprints := getFingerprints()
+
+	thumbprints := []*string{}
+
+	for _, fingerprint := range fingerprints {
+		thumbprints = append(thumbprints, fi.String(fingerprint))
 	}
 
-	if issuerURL == "" {
-		return nil
+	skTask := signingKeyTaskObject.(*fitasks.Keypair)
+
+	keys := &OIDCKeys{
+		SigningKey: skTask,
+	}
+
+	discovery, err := buildDicoveryJSON(serviceAccountIssuer)
+	if err != nil {
+		return err
+	}
+	keysFile := &fitasks.ManagedFile{
+		Contents:  keys,
+		Lifecycle: b.Lifecycle,
+		Location:  fi.String("oidc/keys.json"),
+		Name:      fi.String("keys.json"),
+		Base:      fi.String(b.Cluster.Spec.PublicDataStore),
+		Public:    fi.Bool(true),
+	}
+	c.AddTask(keysFile)
+
+	discoveryFile := &fitasks.ManagedFile{
+		Contents:  fi.NewBytesResource(discovery),
+		Lifecycle: b.Lifecycle,
+		Location:  fi.String("oidc/.well-known/openid-configuration"),
+		Name:      fi.String("discovery.json"),
+		Base:      fi.String(b.Cluster.Spec.PublicDataStore),
+		Public:    fi.Bool(true),
 	}
+	c.AddTask(discoveryFile)
 
 	c.AddTask(&awstasks.IAMOIDCProvider{
 		Name:        fi.String(b.ClusterName()),
 		Lifecycle:   b.Lifecycle,
-		URL:         fi.String(issuerURL),
+		URL:         fi.String(serviceAccountIssuer),
 		ClientIDs:   []*string{fi.String(defaultAudience)},
-		Tags:        b.CloudTags(b.ClusterName(), false),
 		Thumbprints: thumbprints,
 	})
 
 	return nil
 }
+
+func buildDicoveryJSON(issuerURL string) ([]byte, error) {
+	d := oidcDiscovery{
+		Issuer:                fmt.Sprintf("%v/", issuerURL),
+		JWKSURI:               fmt.Sprintf("%v/keys.json", issuerURL),
+		AuthorizationEndpoint: "urn:kubernetes:programmatic_authorization",
+		ResponseTypes:         []string{"id_token"},
+		SubjectTypes:          []string{"public"},
+		SigningAlgs:           []string{"RS256"},
+		ClaimsSupported:       []string{"sub", "iss"},
+	}
+	return json.MarshalIndent(d, "", "")
+}
+
+type KeyResponse struct {
+	Keys []jose.JSONWebKey `json:"keys"`
```

Comment:
Nit: are taking the dependency on the jose library _only_ for the structs?  Should we just embed them?